[PyTorch] RFC: ExclusivelyOwned

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59420 [PyTorch] Try using ExclusivelyOwned in LinearAlgebra
* #59827 gen ExclusivelyOwned in structured kernels
* **#59419 [PyTorch] RFC: ExclusivelyOwned**

This introduces ExclusivelyOwned, which allows isolated
pieces of code that can make ownership guarantees to opt out of
reference counting operations on `intrusive_ptr` and `Tensor`
entirely. To elaborate, if you know you are the exclusive owner of an
`intrusive_ptr` or `Tensor`, moving it into an `ExclusivelyOwned` will
avoid performing atomic reference counting operations at destruction
time. The documentation comment should provide sufficient explanation; please request changes if not.

Differential Revision: [D28885314](https://our.internmc.facebook.com/intern/diff/D28885314/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28885314/)!

swolchok committed:
[PyTorch] RFC: ExclusivelyOwned

This introduces ExclusivelyOwned, which allows isolated
pieces of code that can make ownership guarantees to opt out of
reference counting operations on `intrusive_ptr` and `Tensor`
entirely. To elaborate, if you know you are the exclusive owner of an
`intrusive_ptr` or `Tensor`, moving it into an `ExclusivelyOwned` will
avoid performing atomic reference counting operations at destruction
time. The documentation comment should provide sufficient explanation; please request changes if not.

Differential Revision: [D28885314](https://our.internmc.facebook.com/intern/diff/D28885314/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28885314/)!

[ghstack-poisoned]

swolchok commented:
an open question: would it be nice if Tensor factory functions could optionally return this? We can't simply change the return type because that would make this opt-out instead of opt-in, but perhaps we could have a template wrapper `callOwned<someFactory>()` that had the same signature as someFactory, except that it returned `ExclusivelyOwned<Tensor>`. We could statically check (probably in debug builds only to save resources) that `someFactory` was actually one of the factory functions defined in native_functions.yaml (e.g., by putting their addresses in a constexpr array of `void*`and writing a constexpr function to loop over the array) to prevent misuse.

ezyang commented on aten/src/ATen/templates/TensorBody.h:
```diff
@@ -976,6 +976,57 @@ struct MaybeOwnedTraits<at::Tensor> {
+
+  static repr_type nullRepr() {
+    // REVIEW: do we need to be able to distinguish an empty
+    // ExclusivelyOwned<Tensor> from an undefined Tensor?
```
seems unlikely

ezyang replied:
Alternately, we could delete the default constructor from ExclusivelyOwned (I am not sure how painful that will be for consumers)

ezyang commented on aten/src/ATen/templates/TensorBody.h:
```diff
@@ -976,6 +976,57 @@ struct MaybeOwnedTraits<at::Tensor> {
+#ifndef NDEBUG
+      // Needed to pass the debug assertions in ~intrusive_ptr_target.
+      toDestroy->refcount_ = 0;
+      toDestroy->weakcount_ = 0;
```
Hmm, this is extra invariant on `release_resources` that the refcount/weakcount may not be accurate

swolchok replied:
if release_resources cares about the refcount/weakcount....why?
We can do the stores in all builds if you think it will make a difference to anyone, I would just rather not have the code size and memory traffic.

swolchok replied:
oh, wait, it *can't* care about the refcount & weakcount because they're private to intrusive_ptr_target!

ezyang replied:
I don't think `release_resources` ever relies on it, but it's a near miss, as some PyObject binding stuff uses `use_count` to decide if it has exclusive ownership or not. Just need to doc and socialize this, no action needed.

ezyang commented on c10/util/ExclusivelyOwned.h:
```diff
@@ -0,0 +1,105 @@
+    // End the lifetime of repr_ without executing its dtor, since we
+    // already did specialized destruction for the exclusively-owned
+    // case in destroyOwned!
+    dummy_ = '\0';
```
C unions never destruct anyway, though?

swolchok replied:
normally, we would have to write an explicit destructor call for exactly that reason.

ezyang replied:
I guess my Q is, couldn't we just omit nulling it out here entirely?

swolchok replied:
yes, I think you're right. the lifetime of the union is going to end anyway. will fix.

swolchok committed:
Update on "[PyTorch] RFC: ExclusivelyOwned"


This introduces ExclusivelyOwned, which allows isolated
pieces of code that can make ownership guarantees to opt out of
reference counting operations on `intrusive_ptr` and `Tensor`
entirely. To elaborate, if you know you are the exclusive owner of an
`intrusive_ptr` or `Tensor`, moving it into an `ExclusivelyOwned` will
avoid performing atomic reference counting operations at destruction
time. The documentation comment should provide sufficient explanation; please request changes if not.

Differential Revision: [D28885314](https://our.internmc.facebook.com/intern/diff/D28885314/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28885314/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] RFC: ExclusivelyOwned"


This introduces ExclusivelyOwned, which allows isolated
pieces of code that can make ownership guarantees to opt out of
reference counting operations on `intrusive_ptr` and `Tensor`
entirely. To elaborate, if you know you are the exclusive owner of an
`intrusive_ptr` or `Tensor`, moving it into an `ExclusivelyOwned` will
avoid performing atomic reference counting operations at destruction
time. The documentation comment should provide sufficient explanation; please request changes if not.

Differential Revision: [D28885314](https://our.internmc.facebook.com/intern/diff/D28885314/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28885314/)!

[ghstack-poisoned]

ezyang commented on c10/util/ExclusivelyOwned.h:
```diff
@@ -0,0 +1,123 @@
+// have this traits class, and instead we should just have two
+// explicit specializations of ExclusivelyOwned? There would be some
+// small amount of code duplication, but perhaps it would be easier to
+// understand?
```
I don't have an opinion here, you should do what you feel is best.

ezyang reviewed (approved):
Speculatively approving this. I don't know how this is going to pan out overall in the codebase but it seems like a useful tool to have in the toolbox

swolchok committed:
Update on "[PyTorch] RFC: ExclusivelyOwned"


This introduces ExclusivelyOwned, which allows isolated
pieces of code that can make ownership guarantees to opt out of
reference counting operations on `intrusive_ptr` and `Tensor`
entirely. To elaborate, if you know you are the exclusive owner of an
`intrusive_ptr` or `Tensor`, moving it into an `ExclusivelyOwned` will
avoid performing atomic reference counting operations at destruction
time. The documentation comment should provide sufficient explanation; please request changes if not.

Differential Revision: [D28885314](https://our.internmc.facebook.com/intern/diff/D28885314/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28885314/)!

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] RFC: ExclusivelyOwned"


This introduces ExclusivelyOwned, which allows isolated
pieces of code that can make ownership guarantees to opt out of
reference counting operations on `intrusive_ptr` and `Tensor`
entirely. To elaborate, if you know you are the exclusive owner of an
`intrusive_ptr` or `Tensor`, moving it into an `ExclusivelyOwned` will
avoid performing atomic reference counting operations at destruction
time. The documentation comment should provide sufficient explanation; please request changes if not.

Differential Revision: [D28885314](https://our.internmc.facebook.com/intern/diff/D28885314/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28885314/)!

[ghstack-poisoned]
